src/modules/compliance/src/lib/procedures/EnsureFilePermissions.cpp

```diff
@@ -342,4 +332,121 @@ REMEDIATE_FN(EnsureFilePermissions, "filename:Path to the file:M", "owner:Requir
     OsConfigLogDebug(log, "File '%s' remediation succeeded", filename.c_str());
     return Status::Compliant;
 }
+
+Result<Status> EnsureFilePermissionsCollectionHelper(std::map<std::string, std::string> args, IndicatorsTree& indicators, ContextInterface& context, bool isRemediation)
+{
+    auto log = context.GetLogHandle();
+    auto it = args.find("directory");
+    if (it == args.end())
+    {
+        OsConfigLogError(log, "No directory provided");
+        return Error("No directory provided", EINVAL);
+    }
+    auto directory = std::move(it->second);
+    it = args.find("ext");
+    if (it == args.end())
+    {
+        OsConfigLogError(log, "No file pattern provided");
+        return Error("No file pattern provided", EINVAL);
+    }
+    auto ext = std::move(it->second);
+
+    auto directorycharp = std::unique_ptr<char, void (*)(void*)>(strdup(directory.c_str()), free);
+    if (nullptr == directorycharp)
+    {
+        OsConfigLogError(log, "Failed to allocate memory for directory path");
+        return Error("Failed to allocate memory for directory path", ENOMEM);
+    }
+    char* const paths[] = {directorycharp.get(), nullptr};
+    FTS* ftsp = fts_open(paths, FTS_PHYSICAL | FTS_NOCHDIR, NULL);
+
+    if (!ftsp)
+    {
+        OsConfigLogInfo(log, "Directory '%s' does not exist", directory.c_str());
+        return indicators.Compliant("Directory '" + directory + "' does not exist");
+    }
+    auto ftspDeleter = std::unique_ptr<FTS, int (*)(FTS*)>(ftsp, fts_close);
+    bool hasFiles = false;
+
+    FTSENT* entry = nullptr;
+    while (nullptr != (entry = fts_read(ftsp)))
+    {
+        if (FTS_F == entry->fts_info)
+        {
+            if (0 == fnmatch(ext.c_str(), entry->fts_name, 0))
+            {
+                hasFiles = true;
+                const char* fileName = entry->fts_path;
+
+                Result<Status> result = isRemediation ? RemediateEnsureFilePermissionsHelper(fileName, args, indicators, context) :
+                                                        AuditEnsureFilePermissionsHelper(fileName, args, indicators, context);
```

Comment:
This doesn't have to be acted on now, but we're copying `args` here for every loop iteration. I'm not sure if we're running this for large directories, but consider passing `args` by `const &` to the helpers? I don't think they necessarily need to modify the args.

Reply:
Done.